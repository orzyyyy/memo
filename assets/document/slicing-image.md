<ul>
<li><p>作用</p>
<p>用作 github profile 左下角的 organization icon，就像<a href="https://github.com/orzyyyy">这个</a>一样。</p>
</li>
<li><p>用法</p>
<ol>
<li>默认 7 行 2 列。如有修改，则修安排自行换算下方 organization 数量。</li>
<li>首先你得申请 14 个 organization。以 0- 开头可以防止在加入正经的 organization 时，被展示的 icon 被分割的问题。</li>
<li>上传一张图。</li>
<li>点下载。1-1 表示从左往右、从上到下，第一行第一张，其他的以此类推。</li>
<li>挨个替换 organization 的 icon。</li>
</ol>
</li>
<li><p>流程：</p>
<ol>
<li>把图转成 canvas。</li>
<li>默认以两行七列的规格计算每个 icon 的长宽。</li>
<li>把每个 icon 缩放进 35 * 35 的格子，并渲染到预览框。</li>
<li>为每个 icon 创建 a 标签，触发 click 事件。</li>
</ol>
</li>
</ul>
