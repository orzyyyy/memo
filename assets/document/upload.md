<ul>
<li><p>作用</p>
<p>主要用于局域网内传文件。如果需要部署到服务器上，则需要自行保暖。</p>
</li>
</ul>
