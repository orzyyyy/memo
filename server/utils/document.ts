import fs from 'fs-extra';
import path from 'path';
import { MappingProps } from '../controller/DocumentController';
import { info, error } from './log';
import { joinWithRootPath } from './common';

const checkCategory = (category: string) => {
  if (category !== 'mapping' && category !== 'markdown') {
    throw Error('category is neither "mapping" nor "markdown"');
  }
};

const unique = (target: any[]) => {
  const result: any = {};
  target.forEach(item => {
    result[JSON.stringify(item)] = item;
  });
  return Object.keys(result).map(item => JSON.parse(item));
};

const updateSider = () => {
  const mappingPath = path.join(process.cwd(), 'src/assets/mapping.json');
  const mappingFile = fs.readFileSync(mappingPath);
  const result: MappingProps[] = JSON.parse(mappingFile.toString());

  let types: string[] = [];
  result.map(item => {
    if (item.type) {
      types.push(item.type);
    }
  });
  types = Array.from(new Set(types));

  let menu = [];
  for (const type of types) {
    menu.push({
      key: type,
      title: type,
      children: [] as { key: string; value: string }[],
    });
  }
  for (const item of result) {
    for (const menuItem of menu) {
      if (menuItem.key === item.type) {
        if (item.subType) {
          menuItem.children.push({
            key: item.subType,
            value: item.subType,
          });
        }
      }
    }
  }
  for (const item of menu) {
    item.children = unique(item.children);
  }
  menu = [
    { key: 'all', title: 'all' },
    { key: 'ex-hentai-module', title: 'ex-hentai' },
    ...menu,
  ];
  fs.outputJSON(joinWithRootPath('src/assets/sider.json'), menu, {
    spaces: 2,
  })
    .then(() => {
      info('generate menu-sider completed.');
    })
    .catch(err => {
      error(err);
    });
};

const getWriteFilesPaths = (category?: 'markdown' | 'mapping', id?: string) => {
  if (!id || !category) {
    return [`src/assets/mapping.json`, `dist/assets/mapping.json`];
  }
  checkCategory(category);
  return [
    `src/assets/${category}/${id}.${category === 'mapping' ? 'json' : 'md'}`,
    `dist/assets/${category}/${id}.${category === 'mapping' ? 'json' : 'md'}`,
  ];
};

export { getWriteFilesPaths, updateSider };
