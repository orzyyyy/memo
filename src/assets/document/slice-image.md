- 作用

用作 github profile 左下角的 organization icon，就像[这个](https://github.com/orzyyyy)一样。

- 用法

  1. 首先你得申请 14 个 organization。以 0- 开头可以防止在加入正经的 organization 时，被展示的 icon 被分隔的问题
  2. 上传一张图
  3. 点下载。1-1 表示从左往右、从上到下，第一行第一张，其他的以此类推。
  4. 挨个替换 icon

- 流程：

  1. 把图转成 canvas
  2. 以两行七列的规格计算每个 icon 的长宽
  3. 把每个 icon 缩放进 35 \* 35 的格子，并渲染到预览框
  4. 下载时为每个 icon 创建 a 标签
