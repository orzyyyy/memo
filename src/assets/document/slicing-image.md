- 作用

  用作 github profile 左下角的 organization icon，就像[这个](https://github.com/orzyyyy)一样。

- 用法

  1. 默认 7 行 2 列。如有修改，则修安排自行换算下方 organization 数量。
  2. 首先你得申请 14 个 organization。以 0- 开头可以防止在加入正经的 organization 时，被展示的 icon 被分割的问题。
  3. 上传一张图。
  4. 点下载。1-1 表示从左往右、从上到下，第一行第一张，其他的以此类推。
  5. 挨个替换 organization 的 icon。

- 流程：

  1. 把图转成 canvas。
  2. 默认以两行七列的规格计算每个 icon 的长宽。
  3. 把每个 icon 缩放进 35 \* 35 的格子，并渲染到预览框。
  4. 为每个 icon 创建 a 标签，触发 click 事件。
