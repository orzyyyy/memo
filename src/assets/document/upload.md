主要用于局域网内传文件。如果需要部署到服务器上，则需要自行保暖。
