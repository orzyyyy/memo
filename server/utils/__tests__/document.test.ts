import { updateSider, getWriteMappingPaths } from '../document';
const sider = require('../../../src/assets/sider.json');

describe('util-document', () => {
  it('updateSider', () => {
    expect(updateSider()).toEqual(sider);
  });

  it('getWriteMappingPaths', () => {
    expect(getWriteMappingPaths()).toMatchSnapshot();
    expect(getWriteMappingPaths('markdown', 'test-markdown')).toMatchSnapshot();
    expect(getWriteMappingPaths('mapping', 'test-mapping')).toMatchSnapshot();
  });
});
