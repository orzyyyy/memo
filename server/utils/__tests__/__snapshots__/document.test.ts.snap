// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util-document getWriteMappingPaths 1`] = `
Array [
  "src/assets/mapping.json",
  "dist/assets/mapping.json",
]
`;

exports[`util-document getWriteMappingPaths 2`] = `
Array [
  "src/assets/markdown/test-markdown.md",
  "dist/markdown/test-markdown/test-markdown.md",
  "dist/markdown-editor/test-markdown/test-markdown.md",
]
`;

exports[`util-document getWriteMappingPaths 3`] = `
Array [
  "src/assets/mapping/test-mapping.json",
  "dist/mapping/test-mapping/test-mapping.json",
  "dist/mapping-editor/test-mapping/test-mapping.json",
]
`;
